### PR TITLE
[refactoring] runtime: define and enforce naming conventions for STW functions

### PR DESCRIPTION
**Pull request by gasche**
When reviewing #12597, @kayceesrk asked about our naming conventions around STW sections: is there a conventional suffix for functions that must run within a STW critical section, and where is it documented? ( https://github.com/ocaml/ocaml/pull/12597#discussion_r1340006595 )

The present PR answers this question by defining and documenting a convention around STW sections, and enforcing it in the runtime.

This is purely a documentation and renaming PR, there must be no behavior change. (In the process I found a few places where the code could profitably be changed, they are marked as TODO items.)

To review this PR, one should start by looking at the domain.h diff which specifies the naming conventions, before looking at runtime changes enforcing the conventions or commenting on their use.

The gist of the naming convention proposed:

- use `stw_foo` or `caml_stw_foo` for whole STW callbacks (there was another convention in play in some parts of the runtime to use `foo_callback`, those are renamed accordingly)
- use `foo_from_stw` for functions that can only be called within a STW critical section (when mutators cannot run)
- use `foo_from_stw_single` for functions that must be called from a single domain within a STW critical section (typically, the last one entering a barrier)

Remark: another refactoring PR I submitted based on KC's feedback was #12617. For that PR I was not sure that the benefits of the PR are worth the code churn. In the present case I feel fairly confident that this PR is a desirable change, because it enforces (through names) a discipline on the most delicate synchronization mechanism in the runtime.

Anyone interested in the concurrent runtime code could review this PR. (For example @kayceesrk, @gadmm, @Engil, @fabbing, @OlivierNicole, etc. )

### CODE REVIEW

File: runtime/domain.c
Comment (abbysmal) [inline runtime/domain.c]: `reserve_minor_heaps`'s name was not changed here

File: runtime/runtime_events.c
Comment (abbysmal) [inline runtime/runtime_events.c]: Just to check I get this right: it is not required because it is called from a STW segment and the barrier is therefore not needed? (since it runs from a single domain that is already checked for in the next block?)

File: runtime/caml/domain.h
Comment (abbysmal) [inline runtime/caml/domain.h:160]: Question: should we find a naming convention for functions starting a STW segment? Not obvious to me if the clarification is needed, but it does kind of sound more organic to me.

**Review by abbysmal — COMMENTED** [review]
I think the naming convention is a welcome addition and make it clearer what part of a stop the world segment we are looking at.

I made a few comments, I think the documentation for the convention itself is fine.

One last thing: I think the `stw_single` terminology for does not sound exactly right. I'm think `stw_single_domain` may make it a little bit more obvious at first sight, but then again the documentation is pretty clear on the intent so this is purely bike shedding.

File: runtime/runtime_events.c
Comment (gasche) [reply runtime/runtime_events.c]: My reasoning is the following.

STW callbacks only start running when all registered domains have synchronized, which amounts to a barrier -- you can see barrier-looking code in at the beginning of `stw_handler` in domain.c; for the leader the barrier behavior comes from the calls to `wait_interrupt_serviced`.

Additional/explicit barriers are only useful in the middle or at the end of a STW callback, to guarantee that the callback logic before the barrier does not race with the callback logic after the barrier or (at the end) with mutators resuming after the STW callback.

File: runtime/caml/domain.h
Comment (gasche) [reply runtime/caml/domain.h:160]: My idea is that the naming convention helps when functions may only be called safely in certain contexts. For me, calling a STW pause is an implementation detail. For example a given runtime service may be implemented by either taking a lock on both reads and writes to an internal global resource, or by calling a STW pause on writes and doing nothing on reads. It makes sense to use the same function names for either implementations.

(For example in signals.c, `caml_install_signal_handler` uses a lock to protect a global variable; we might in the future want to update it to calling a STW section internally to avoid a race with `caml_execute_signal_exn`. But this needs not change how the callers refer to the function.)

Technically, calling a STW pause (directly or transitively) does restrict the contexts in which one can run, as STW pauses cannot be nested. This is in fact similar to taking locks, which also cannot be nested. But trying to enforce this discipline through naming would basically amount to marking half the runtime functions as "toplevel" functions and the other half as "leaf" functions, this would be invasive and not very useful.

File: runtime/domain.c
Comment (gasche) [reply runtime/domain.c]: Fixed, thanks! (In the logs the un-suffixed names are used in places, but I prefer to keep them as-is as they are shorter and introduce no risk of confusion.)

**Comment by gasche** [comment]
( @Engil when you have time, I would be curious to know if my answer were convincing, and I wonder if you are planning to approve the PR or if there are other changes to consider / you think someone else should have a look. )

**Review by ghost — APPROVED** [review]

**Comment by gasche** [comment]
Thanks! This PR has been reviewed by @Engil and @dustanddreams and approved by @dustanddreams. It needs a maintainer approval, either after an independent review or on behalf of @dustanddreams.

( cc @Engil, @Octachron , maybe @nojb? )

**Comment by gasche** [comment]
I guess that @eutro could also be interested in this change, because it interacts with #12579 somewhat; there are minor conflicts (but conflicts related to naming-only changes tend to be easy to resolve) and the discussion of barriers and critical sections may overlap with some of the documentation or refactoring ideas in #12579. (I am not suggesting that @eutro should review, I think we have enough reviews for a naming-only change by now.)

File: runtime/runtime_events.c
Comment (eutro) [reply runtime/runtime_events.c]: I don't believe this is correct. The STW code uses the runtime events API between the "enter barrier" and the time the domain starts running the STW section [here](https://github.com/ocaml/ocaml/blob/a912cff51de76b64c8341e39e4ec60361724553b/runtime/domain.c#L1345). That is, it's not mutator code we'd be "racing" with, but the runtime there. Removing this barrier coincidentally wouldn't be a data race, I believe, but it would mean that orphan `END` events could be emitted on the line I linked, depending on how the threads happen to be scheduled. Removing the other similar double barrier in the teardown STW does cause a race and horrible things.

It should be sufficient to have one single barrier use, running the setup/teardown from the final party while all others are blocked waiting for the release. In #12579 I do this and add a comment, similar to my explanation above, detailing why it has to be done this way.

**Review by eutro — CHANGES_REQUESTED** [review]
I don't believe the TODOs in `stw_(create|teardown)_runtime_events` are correct.

File: runtime/major_gc.c
Comment (eutro) [inline runtime/major_gc.c]: Are you sure that `caml_verify_heap_from_stw` wouldn't race with the GC code below? I'm sceptical since they both access global roots, for example.

File: runtime/runtime_events.c
Comment (gasche) [reply runtime/runtime_events.c]: Thanks for your comments, I had indeed not considered the case of races between the code here and the instruction inside the STW runner functions. I will remove the offending TODOs and leave any change there to your PR. (In particular, I am happy to delegate to you the responsibility of document the subtle thing that is going on there.)

File: runtime/major_gc.c
Comment (gasche) [reply runtime/major_gc.c]: I removed the TODO and added your explanation in a comment. Thanks!

**Comment by gasche** [comment]
I took @eutro's comments into account.

**Review by eutro — APPROVED** [review]

File: runtime/caml/domain.h
Comment (NickBarnes) [inline runtime/caml/domain.h]: I suggest "must only be called" or "may only be called".

File: runtime/caml/domain.h
Comment (NickBarnes) [inline runtime/caml/domain.h]: I suggest "may only be called within an STW critical section, and only by a single domain at a time --". When thinking about this issue during my review of #12579 I considered `_in_stw_solo` where you have `_from_stw_single`, but obviously either is fine.

**Review by NickBarnes — APPROVED** [review]
Good.

**Review by Octachron — APPROVED** [review]
Let me green stamp the three existing reviews (by people who have worked on the area of interest) on this naming convention change.

File: runtime/caml/domain.h
Comment (gasche) [reply runtime/caml/domain.h]: Thanks!

**Comment by gasche** [comment]
Thanks everyone. This is the most heavily reviewed renaming PR in the history of this academic year, and the result will be all the better.

I am planning to merge if the CI passes. (This will push some rebase work onto @eutro for #12579, but hopefully this will be easy and might even clarify some things.)

**Comment by NickBarnes** [comment]
It's easy to laugh at the effort put into something like this, but naming is important and surprisingly difficult!

**Pull request merged**
